test/managers/integration/MerkleProofManager.t.sol

```diff
@@ -0,0 +1,511 @@
+            UpdateRestrictionMessageLib.UpdateRestrictionMember({
+                user: address(this).toBytes32(),
+                validUntil: MAX_UINT64
+            }).serialize()
+        );
+
+        manager = new MerkleProofManager(POOL_A, address(spoke));
+    }
+
+    function _depositIntoBalanceSheet(uint128 amount) internal {
+        erc20.mint(address(this), amount);
+        erc20.approve(address(balanceSheet), amount);
+        balanceSheet.deposit(POOL_A, defaultTypedShareClassId, address(erc20), erc20TokenId, amount);
+    }
+
+    function _selector(string memory signature) internal pure returns (bytes4) {
+        return bytes4(keccak256(abi.encodePacked(signature)));
+    }
+
+    function _computeHashes(PolicyLeaf[] memory policyLeafs) internal pure returns (bytes32[] memory) {
+        bytes32[] memory leafs = new bytes32[](policyLeafs.length);
+        for (uint256 i; i < policyLeafs.length; ++i) {
+            leafs[i] = keccak256(
+                abi.encodePacked(
+                    policyLeafs[i].decoder,
+                    policyLeafs[i].target,
+                    policyLeafs[i].valueNonZero,
+                    policyLeafs[i].selector,
+                    policyLeafs[i].addresses
+                )
+            );
+        }
+        return leafs;
+    }
+}
+
+contract MerkleProofManagerFailureTests is MerkleProofManagerBaseTest {
+    using CastLib for *;
+    using UpdateContractMessageLib for *;
+
+    function testNotStrategist() public {
+        Call[] memory calls = new Call[](0);
+
+        vm.expectRevert(IMerkleProofManager.NotAStrategist.selector);
+        manager.execute(calls);
+    }
+
+    function testInvalidDecoder() public {
+        decoder = new VaultDecoder();
+        uint128 withdrawAmount = 100_000;
+        _depositIntoBalanceSheet(withdrawAmount);
+
+        // Generate policy root hash
+        PolicyLeaf[] memory leafs = new PolicyLeaf[](2);
+        leafs[0] = PolicyLeaf({
+            decoder: address(decoder),
+            target: address(balanceSheet),
+            valueNonZero: false,
+            selector: _selector("withdraw(uint64,bytes16,address,uint256,address,uint128)"),
+            addresses: abi.encodePacked(erc20, manager)
+        });
+
+        leafs[1] = PolicyLeaf({
+            decoder: address(decoder),
+            target: address(balanceSheet),
+            valueNonZero: false,
+            selector: _selector("deposit(uint64,bytes16,address,uint256,uint128)"),
+            addresses: abi.encodePacked(erc20)
+        });
+
+        bytes32[][] memory tree = MerkleTreeLib.generateMerkleTree(_computeHashes(leafs));
+
+        vm.prank(address(spoke));
+        manager.update(
+            POOL_A,
+            defaultTypedShareClassId,
+            UpdateContractMessageLib.UpdateContractPolicy({
+                who: address(this).toBytes32(),
+                what: tree[tree.length - 1][0]
+            }).serialize()
+        );
```

Comment:
Nit: This can be moved to a helper function as its used 4 times throughout the tests. Would make it much easier to follow.